### PR TITLE
2017-12-18-Ansible-Filter-Plugins.md

### PR DESCRIPTION
**Pull request by mattsonjoshcf**
In a comment on the pull request, include the following information and check off the list.

 - [X] Mini bio
  Senior Application Support Engineer and 8 year Racker on the Rackspace Application Services team. Specializing in Java web applications
 - [X] Desired date of publishing
  12/19/17
 - [X] Social media contact info
   Twitter: @mattsonjoshcf
 - [X] Redaction: (yes|no)
  - Sure
 - [X) Comments: no

@catlook Please review.

### CODE REVIEW

**Comment by rackernexus** [comment]
Your content preview [is now ready](https://staging.developer.rackspace.com/build-3223366dd0/). :bow:

**Comment by rackernexus** [comment]
HULK [PREVIEW](https://staging.developer.rackspace.com/build-5d6b519e02/)

**Comment by rackernexus** [comment]
:eyes: [Preview](https://staging.developer.rackspace.com/build-5d6b519e02/)

**Comment by mattsonjoshcf** [comment]
Sorry for the commit spam.  Apparently `{{ }}` and everything inside gets stripped out when its rendered on the page.  Maybe Jinja is in use for the blog?  I believe I may have it resolved now, but I"m not entirely certain as that last preview didn't update after the final commit. In any case, if it's still not displaying correctly, I'm all ears for how to correct this.

**Comment by catlook** [comment]
I'll review and edit!

File: _posts/2017-12-18-Ansible-Filter-Plugins.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+comments: false
+author: Josh Mattson
+published: true
+authorIsRacker: true
+authorAvatar: 'http://www.gravatar.com/avatar/791617263c70278859e1b26c15d13eab'
+categories:
+    - DevOps
+---
+
+[Rackspace Application Services](https://www.rackspace.com/en-us/digital/rackspace-application-services) provides application support and management to a wide variety of customers ranging in size from small environments with only a few application servers to customers that run thousands of JVMs across their environment.  To help facilitate this, we heavily rely on [Ansible](https://www.ansible.com/) to help us automate implementation, troubleshooting and maintenance tasks.  While Ansible is quite powerful and easy to use, many organizations do not take full advantage of some of the features that it provides.  In this article, we'll be discussing how you can extend Jinja2 and Ansible's built-in [filter plugins](http://docs.ansible.com/ansible/latest/dev_guide/developing_plugins.html#filter-plugins) and how you can craft a completely new filter plugin to make specific tasks easier.
```
Comment (catlook) [inline _posts/2017-12-18-Ansible-Filter-Plugins.md]: ... thousands of Java Virtual Machines (or JVMs) across their environment ...
... troubleshooting, and maintenance ...

File: _posts/2017-12-18-Ansible-Filter-Plugins.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
+---
+layout: post
+title: "Ansible Filter Plugins"
+date: 2017-12-18 00:00
```
Comment (catlook) [inline _posts/2017-12-18-Ansible-Filter-Plugins.md]: go ahead and change to today's date.  sorry I didn't get to this faster!!

File: _posts/2017-12-18-Ansible-Filter-Plugins.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+<!-- more -->
+
+### Filters: The Basics
+
+Jinja2 comes with a [significant number](http://jinja.pocoo.org/docs/2.10/templates/#builtin-filters) of filters that are, at their core, simply a way to transform data.  Ansible has also extended and expanded upon these filters to perform more common tasks related to systems orchestration & management.
```
Comment (catlook) [inline _posts/2017-12-18-Ansible-Filter-Plugins.md]: ... orchestration and management.

File: _posts/2017-12-18-Ansible-Filter-Plugins.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+Let's look at a very simple example to illustrate the concept.  Suppose that we have a variable that can be inputted in to one of our playbooks, but this variable may contain leading or trailing whitespace which your playbook needs to account for – perhaps an API key or something that could easily run in to issues if incorrectly copy & pasted when executing the playbook.  We can use the built-in Jinja2 filter `trim` to address this in our template to very quickly strip out all extraneous whitespace:
+
+
+`api_key: "{{ api_key | trim }}"`
+
+This is obviously a very simple example, but it should illustrate the power of filter plugins.
```
Comment (catlook) [inline _posts/2017-12-18-Ansible-Filter-Plugins.md]: ... but it illustrates the power ...

File: _posts/2017-12-18-Ansible-Filter-Plugins.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+### Creating Custom Filter Plugins
+
+The built-in plugins within Jinja2 and Ansible are all well and good, but we've found situations where we have a much more specific use case.  To use a real world example, we maintain several thousand APM agents for our customers environments, and it can be a challenge to ensure that we are providing timely updates to these agents while being cognizant of customer preferences for timing, change control, etc.
```
Comment (catlook) [inline _posts/2017-12-18-Ansible-Filter-Plugins.md]: .... timing, change control, and so on.

File: _posts/2017-12-18-Ansible-Filter-Plugins.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+With these specific agents, the best way to determine the version of the existing agent is to make an API call to the controller and request the agent version.  However, the API will return a string similar to this:
+
+`Server Agent v4.2.0.0 GA #10348 r6cb1b69515a61e8577749ec847709720770f46f3 31-4.2.0.next-build`
+
+Ultimately when we're recording Ansible facts and using these to determine which agents need to be updated, we only care about the actual version number itself.  Enter a bit of Python for the custom filter plugin which we simply drop in to the `filter_plugins/` directory in the root of your Ansible repository:
```
Comment (catlook) [inline _posts/2017-12-18-Ansible-Filter-Plugins.md]: Ultimately, when ....

... plugin, which ...

File: _posts/2017-12-18-Ansible-Filter-Plugins.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        version_search = re.search(r'v(\d[^\s]+)', string, re.M|re.I)
+        if not version_search:
+            raise AppDynamicsVersionFormat("{} isn't a valid AppDynamics version string.".format(string))
+        return version_search.group(1)
+
+class AppDynamicsVersionFormat(Exception):
+    pass
+```
+
+If you're unfamiliar with Python, essentially all we're doing here is declaring a new FilterModule class that performs a simple regex evaluation on a string and, assuming it is correctly recognized as a valid AppDynamics agent version string, returns our nicely formatted version string.
```
Comment (catlook) [inline _posts/2017-12-18-Ansible-Filter-Plugins.md]: ... Python, all ...

File: _posts/2017-12-18-Ansible-Filter-Plugins.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+To see our custom filter plugin in action, here is how we utilize it in our playbook.  After making our API call to get all agent versions and storing it in a dictionary, we can simply iterate through the dictionary for each agent to grab the raw output and pass it through this filter just as if we were using a built-in filter:
```
Comment (catlook) [inline _posts/2017-12-18-Ansible-Filter-Plugins.md]: ... we use it ...
... we  iterate through ...

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-f95455abc9/).

**Pull request merged**